velox/type/StringViewIdMap.h

```diff
@@ -0,0 +1,274 @@
+    return id;
+  }
+
+#define PREPARE1(n)  \
+  int32_t offset##n; \
+  int64_t head##n;   \
+  prepare1(table, &views[indices[n]], head##n, offset##n);
+
+#define PROBE1(n, make)   \
+  probe1<make>(           \
+      table,              \
+      &views[indices[n]], \
+      head##n,            \
+      offset##n,          \
+      ids[n],             \
+      extraData,          \
+      tailWriter);
+
+  /// Does a batch of 8 findIds on StringViews in an array. The
+  /// indices are in 'indices' and the ids are returned in
+  /// 'ids'. Other parameters are as in findId(). This is noticeably
+  /// faster if data is not all in cache.
+  template <bool makeId>
+  void findIds8(
+      const StringView* views,
+      const int32_t* indices,
+      int32_t* ids,
+      void* extraData = nullptr,
+      StringCopyHook tailWriter = nullptr) {
+    if (makeId && numEntries_ + 8 > maxEntries_) {
+      resize(capacity_ * 2);
+    }
+    auto* table = table_.data();
+    PREPARE1(0);
+    PREPARE1(1);
+    PREPARE1(2);
+    PREPARE1(3);
+    PREPARE1(4);
+    PREPARE1(5);
+    PREPARE1(6);
+    PREPARE1(7);
+    PROBE1(0, makeId);
+    PROBE1(1, makeId);
+    PROBE1(2, makeId);
+    PROBE1(3, makeId);
+    PROBE1(4, makeId);
+    PROBE1(5, makeId);
+    PROBE1(6, makeId);
+    PROBE1(7, makeId);
+  }
+
+#undef PREPARE
+#undef PROBE
+
+  void clear();
+
+ private:
+  static constexpr int32_t kNoEmpty = ~0;
+  static constexpr int64_t kEmpty = 0;
+  static constexpr int32_t kEntrySize = 20;
+
+  int32_t startOffset(uint32_t seed, uint64_t data) {
+    return (simd::crc32U64(seed, data) & sizeMask_) * kEntrySize;
+  }
+
+  template <typename T>
+  T& itemAt(void* table, int32_t offset) {
+    return *reinterpret_cast<T*>(reinterpret_cast<char*>(table) + offset);
+  }
+
+  template <typename T>
+  T itemAt(const void* table, int32_t offset) {
+    return *reinterpret_cast<const T*>(
+        reinterpret_cast<const char*>(table) + offset);
+  }
+
+  FOLLY_ALWAYS_INLINE void prepare1(
+      uint8_t* table,
+      const StringView* view,
+      int64_t& head,
+      int32_t& offset) {
+    head = view->sizeAndPrefixAsInt64();
+    auto size = static_cast<uint32_t>(head);
+    if (size <= 4) {
+      offset = startOffset(1, head);
+    } else if (size <= 12) {
+      auto tail = itemAt<int64_t>(view, 8);
+      offset = startOffset(head >> 32, tail);
+    } else {
+      auto string = itemAt<char*>(view, 8);
+      auto last = itemAt<int64_t>(string, size - 8);
+      offset = startOffset(head >> 32, last);
+    }
+    __builtin_prefetch(&itemAt<int64_t>(table, offset));
+  }
+
+  template <bool makeId = true>
+  FOLLY_ALWAYS_INLINE void probe1(
+      uint8_t* table,
+      const StringView* view,
+      int64_t head,
+      int32_t offset,
+      int32_t& id,
+      void* extraData,
+      StringCopyHook tailWriter) {
+    auto size = static_cast<uint32_t>(head);
+    if (size <= 4) {
+      for (;;) {
```

Comment:
We need to track the number of probes and signal it if it becomes too long, since strings with same size and same prefix 4 and suffix 8 bytes will be sent to the same start offset.